### PR TITLE
SGD

### PR DESCRIPTION
**Pull request by yura-ata**
added momentum and Nesterov accelerated gradient to vanilla SGD optimizer
though I'm not 100% sure in how state is handled and whether nesterov formulae is correct (shoould we +/- when maximizing/minimizing fn)

### CODE REVIEW

File: src/main/scala/org/scanet/optimizers/Algorithm.scala [hunk 1]
```diff
@@ -1,10 +1,11 @@
 package org.scanet.optimizers
 
-import org.scanet.core.{Output, TensorType}
+import org.scanet.core.{Output, Tensor, TensorType}
 import org.scanet.math.Numeric
 
-trait Algorithm {
+trait Algorithm[S] {
```
Comment (pashashiz) [inline src/main/scala/org/scanet/optimizers/Algorithm.scala:6]: I think we can avoid having a type param here, what if we just assume that is Float`?

File: src/main/scala/org/scanet/optimizers/SGD.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    // Nesterov == true:
+    //   v[i] = momentum * v[i-1] + rate * grad(arg - momentum * v[i-1])
+    // Nesterov == false:
+    //   v[i] = momentum * v[i-1] + rate * grad(arg)
+    // momentum == 0.0 (vanilla SGD):
+    //   v[i] = rate * grad(arg)
+    // arg = arg - v[i]
+
+    val m = momentum.toFloat.const
+    // TODO: should this be + for maximize?
```
Comment (pashashiz) [inline src/main/scala/org/scanet/optimizers/SGD.scala]: Calculating Nesterow with formula from the paper is problematic and will not allow optimizations in the future. What if we take an implementation from Keras where (see https://arxiv.org/pdf/1212.0901v2.pdf):
```
for p, g, m in zip(params, grads, moments):
            v = self.momentum * m - lr * g  # velocity
            self.updates.append(K.update(m, v))

            if self.nesterov:
                new_p = p + self.momentum * v - lr * g
            else:
                new_p = p + v
```

Comment (pashashiz) [reply src/main/scala/org/scanet/optimizers/SGD.scala]: Also see https://keras.io/api/optimizers/sgd/#sgd-class

File: src/main/scala/org/scanet/optimizers/Algorithm.scala (hunk 1 above)
Comment (yura-ata) [reply src/main/scala/org/scanet/optimizers/Algorithm.scala:6]: I've looked at other optimizers and thought that sometimes we would need pair a pair of tensors as "state". If this is not the case - it would be nice to remove type param

Comment (pashashiz) [reply src/main/scala/org/scanet/optimizers/Algorithm.scala:6]: makes sense

File: src/main/scala/org/scanet/optimizers/SGD.scala [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import org.scanet.math.syntax._
+
+case class SGD(rate: Double = 0.01, momentum: Double = 0.0, nesterov: Boolean = false) extends Algorithm[Tensor[Float]] {
+
+  override def delta[A: Numeric : TensorType, B: Numeric : TensorType](f: TensorFunction[A, B], initArg: Output[A], prevState: Tensor[Float]): (Output[Float], Tensor[Float]) = {
+    val g = f.grad(initArg)
+    val m = momentum.toFloat.const
+    val r = rate.toFloat.const
+    val v = r * g - m * prevState.const
+    val velocity = if (nesterov) r * g - m * v else v
```
Comment (pashashiz) [inline src/main/scala/org/scanet/optimizers/SGD.scala]: I would call it `delta`

File: src/main/scala/org/scanet/optimizers/SGD.scala [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    val res = velocity.eval
+    (res.const, res)
```
Comment (pashashiz) [inline src/main/scala/org/scanet/optimizers/SGD.scala]: I guess the state is a `v` and not `velocity`

Comment (yura-ata) [reply src/main/scala/org/scanet/optimizers/SGD.scala]: thanks! fixed

File: src/main/scala/org/scanet/optimizers/SGD.scala (hunk 3 above)
Comment (yura-ata) [reply src/main/scala/org/scanet/optimizers/SGD.scala]: thanks, improved naming
